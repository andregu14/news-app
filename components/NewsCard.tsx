import { View, Text, ViewProps } from "./Themed";
import { StyleSheet, useColorScheme, Pressable } from "react-native";
import Colors from "@/constants/Colors";
import { Image } from "expo-image";
import { MaterialCommunityIcons } from "@expo/vector-icons";

type NewsCardProps = ViewProps & {
  title: string;
  bodyText: string;
  image: string;
  department: string;
  time: string;
  onPress: () => void;
  testID?: string;
  imageTestID?: string;
};

export default function NewsCard({
  title,
  bodyText,
  image,
  department,
  time,
  style,
  onPress,
  testID,
  imageTestID,
  ...otherProps
}: NewsCardProps) {
  const colorScheme = useColorScheme() ?? "light";
  const themeColors = Colors[colorScheme];

  const timeText = `${
    time[0].toUpperCase() + time.slice(1)
  } • Em ${department}`;
  const accessibilityLabel = `Notícia: ${title}. ${bodyText}. Publicado ${timeText}`;

  return (
    <Pressable
      onPress={onPress}
      testID={testID}
      style={[styles.container, style, {borderColor: themeColors.borderColor}]}
      accessible={true}
      accessibilityRole="button"
      accessibilityLabel={accessibilityLabel}
      accessibilityHint="Toque para ver mais detalhes"
      hitSlop={{ top: 12, bottom: 12, left: 8, right: 8 }}
      {...otherProps}
    >
      <Text style={styles.title} accessible={false} accessibilityRole="header">
        {title}
      </Text>
      <Image
        style={styles.image}
        source={image}
        contentFit="cover"
        testID={imageTestID || "news-card-image"}
        accessible={false}
        accessibilityElementsHidden={true}
      />
      <Text
        style={[styles.bodyText, { color: themeColors.bodyText }]}
        ellipsizeMode={"tail"}
        numberOfLines={4}
        accessible={false}
      >
        {bodyText}
      </Text>

      <View style={styles.footerContainer}>
  <MaterialCommunityIcons name="clock-outline" size={14} color={themeColors.secondaryText} />
  <Text style={[styles.footerText, { color: themeColors.secondaryText, marginLeft: 5 }]}>
    {time[0].toUpperCase() + time.slice(1)}
  </Text>
  <Text style={[styles.footerText, { color: themeColors.secondaryText }]}> • </Text>
  <MaterialCommunityIcons name="tag-outline" size={14} color={themeColors.secondaryText} />
  <Text style={[styles.footerText, { color: themeColors.secondaryText, marginLeft: 5 }]}>
    Em {department}
  </Text>
</View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  container: {
    marginVertical: 20,
    paddingVertical: 20,
    minHeight: 44,
  },
  title: {
    marginLeft: 15,
    marginBottom: 10,
    fontSize: 18,
    fontWeight: "bold",
    lineHeight: 24,
  },
  bodyText: {
    marginVertical: 10,
    paddingHorizontal: 15,
    fontSize: 14,
    lineHeight: 22,
  },
  image: {
    width: "100%",
    height: 180,
    marginTop: 10,
  },
  footerContainer: {
  flexDirection: 'row',
  alignItems: 'center',
  marginTop: 10,
  marginLeft: 15,
},
footerText: {
  fontSize: 12,
  lineHeight: 16,
},
});
