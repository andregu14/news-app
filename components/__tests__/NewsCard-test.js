import { render, fireEvent, screen } from '@testing-library/react-native';
import NewsCard from '../NewsCard';

// Mock do useColorScheme para controlar tema
const mockUseColorScheme = jest.fn(() => 'light');

jest.mock('react-native/Libraries/Utilities/useColorScheme', () => ({
    default: mockUseColorScheme,
}));

// Mock do Colors e departmentColors, se necessário
jest.mock('@/constants/Colors', () => ({
    __esModule: true,
    default: {
        light: {
            text: '#000',
            borderColor: '#ccc',
            subText: '#666',
            cardBackground: '#fff',
            tint: '#007AFF',
        },
        dark: {
            text: '#fff',
            borderColor: '#444',
            subText: '#aaa',
            cardBackground: '#181A20',
            tint: '#0A84FF',
        },
    },
    departmentColors: {}, // Se usar badges, defina aqui
}));

describe('NewsCard', () => {
    beforeEach(() => {
        mockUseColorScheme.mockReturnValue('light');
    });

    it('renders correctly with all props and matches snapshot', () => {
        const testTitle = "Test Title";
        const testBodyText = "This is a test body text.";
        const testDepartment = "Sports";
        const testTime = "há 7 dias";
        const testImage = "https://example.com/test-image.png";

        const component = render(
            <NewsCard
                title={testTitle}
                bodyText={testBodyText}
                department={testDepartment}
                time={testTime}
                image={testImage}
                onPress={() => { }}
                testID="news-card"
                imageTestID="news-card-image"
            />
        );

        // Verifica textos renderizados
        expect(screen.getByText(testTitle)).toBeTruthy();
        expect(screen.getByText(testBodyText)).toBeTruthy();
        expect(
            screen.getByText("Há 7 dias")
        ).toBeTruthy();

        const snapshot = component.toJSON();
        expect(snapshot).toMatchSnapshot();

        const snapshotString = JSON.stringify(snapshot);
        expect(snapshotString).toContain('news-card-image');
        expect(snapshotString).toContain(testImage);
    });

    it('calls onPress prop when pressed', () => {
        const onPressMock = jest.fn();

        render(
            <NewsCard
                title="Title"
                bodyText="Body"
                department="Sports"
                time="há 7 dias"
                image="https://example.com/image.png"
                onPress={onPressMock}
                testID="news-card"
            />
        );

        const card = screen.getByTestId('news-card');
        fireEvent.press(card);

        expect(onPressMock).toHaveBeenCalledTimes(1);
    });
});

describe('Accessibility', () => {
    it('has correct accessibility properties', () => {
        render(
            <NewsCard
                title="Title"
                bodyText="Body"
                department="Sports"
                time="há 7 dias"
                image="https://example.com/image.png"
                onPress={() => { }}
                testID="news-card"
            />
        );

        const card = screen.getByTestId('news-card');
        expect(card.props.accessible).toBe(true);
        expect(card.props.accessibilityRole).toBe('button');
        expect(card.props.accessibilityLabel).toContain('Notícia');
        expect(card.props.accessibilityHint).toBe('Toque para ver mais detalhes');
    });

    it('provides meaningful content descriptions', () => {
        const customTitle = "Custom Title";
        const customBody = "Custom body text";

        render(
            <NewsCard
                title={customTitle}
                bodyText={customBody}
                department="Sports"
                time="há 7 dias"
                image="https://example.com/image.png"
                onPress={() => { }}
            />
        );

        expect(screen.getByText(customTitle)).toBeTruthy();
        expect(screen.getByText(customBody)).toBeTruthy();
        expect(screen.getByText(/há 7 dias/i)).toBeTruthy();
        expect(screen.getByText(/sports/i)).toBeTruthy();
    });
});
