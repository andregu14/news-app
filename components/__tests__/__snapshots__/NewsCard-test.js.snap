// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsCard renders correctly with all props and matches snapshot 1`] = `
<View
  accessibilityHint="Toque para ver mais detalhes"
  accessibilityLabel="Notícia: Test Title. This is a test body text.. Publicado Há 7 dias • Em Sports"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    {
      "bottom": 12,
      "left": 8,
      "right": 8,
      "top": 12,
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "marginVertical": 20,
        "minHeight": 44,
        "paddingVertical": 20,
      },
      undefined,
      {
        "borderColor": "#ccc",
      },
    ]
  }
  testID="news-card"
>
  <Text
    accessibilityRole="header"
    accessible={false}
    style={
      [
        {
          "color": "#000",
          "fontFamily": "Inter_600SemiBold",
          "fontSize": 18,
        },
        {
          "lineHeight": 24,
          "marginBottom": 10,
          "marginLeft": 15,
        },
      ]
    }
  >
    Test Title
  </Text>
  <ViewManagerAdapter_ExpoImage
    accessibilityElementsHidden={true}
    accessible={false}
    containerViewRef={"[React.ref]"}
    contentFit="cover"
    contentPosition={
      {
        "left": "50%",
        "top": "50%",
      }
    }
    height={180}
    marginTop={10}
    nativeViewRef={"[React.ref]"}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    placeholder={[]}
    source={
      [
        {
          "uri": "https://example.com/test-image.png",
        },
      ]
    }
    style={
      {
        "height": 180,
        "marginTop": 10,
        "width": "100%",
      }
    }
    testID="news-card-image"
    transition={null}
    width="100%"
  />
  <Text
    accessible={false}
    ellipsizeMode="tail"
    numberOfLines={4}
    style={
      [
        {
          "color": "#000",
          "fontFamily": "Inter_400Regular",
          "fontSize": 14,
        },
        [
          {
            "lineHeight": 22,
            "marginVertical": 10,
            "paddingHorizontal": 15,
          },
          {
            "color": undefined,
          },
        ],
      ]
    }
  >
    This is a test body text.
  </Text>
  <View
    style={
      [
        {
          "backgroundColor": undefined,
        },
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginLeft": 15,
          "marginTop": 10,
        },
      ]
    }
  >
    <Text />
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "Inter_400Regular",
          },
          [
            {
              "fontSize": 12,
              "lineHeight": 16,
            },
            {
              "color": undefined,
              "marginLeft": 5,
            },
          ],
        ]
      }
    >
      Há 7 dias
    </Text>
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "Inter_400Regular",
          },
          [
            {
              "fontSize": 12,
              "lineHeight": 16,
            },
            {
              "color": undefined,
            },
          ],
        ]
      }
    >
       • 
    </Text>
    <Text />
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "Inter_400Regular",
          },
          [
            {
              "fontSize": 12,
              "lineHeight": 16,
            },
            {
              "color": undefined,
              "marginLeft": 5,
            },
          ],
        ]
      }
    >
      Em 
      Sports
    </Text>
  </View>
</View>
`;
